src/controllers/product.ts

```diff
@@ -0,0 +1,148 @@
+import { Brand } from '@/models/Brand';
+import { Product } from '@/models/Product';
+import { RequestHandler, Request, Response } from 'express';
+import { ParsedQs } from 'qs';
+const { Op } = require('sequelize');
+
+export const search: RequestHandler = async (req, res) => {
+  const { keyword } = req.query;
+
+  const products = await Product.findAll({
+    include: [
+      {
+        model: Brand,
+      },
+    ],
+    where: {
+      [Op.or]: [
+        { name: { [Op.like]: `%${keyword}%` } },
+        { '$brand.name$': { [Op.like]: `%${keyword}%` } },
+      ],
+    },
+  });
+
+  try {
+    return res.status(200).json(products);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+export const getLimitedEdition: RequestHandler = async (req, res) => {
+  req.query.quantity = '20';
+  const filter = createFilter(req.query);
+  try {
+    const result = await getProductsByFilter(filter);
+    return res.status(200).json(result);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+export const Popular: RequestHandler = async (req, res) => {
+  req.query.rating = '4.5';
+  const filter = createFilter(req.query);
+  try {
+    const result = await getProductsByFilter(filter);
+    return res.status(200).json(result);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+export const newArrivals: RequestHandler = async (req, res) => {
+  req.query.isNew = '1';
+  const filter = createFilter(req.query);
+  try {
+    const result = await getProductsByFilter(filter);
+    return res.status(200).json(result);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+export const Handpicked: RequestHandler = async (req, res) => {
+  req.query.handpicked = '1';
+  const filter = createFilter(req.query);
+  try {
+    const result = await getProductsByFilter(filter);
+    return res.status(200).json(result);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+export const getProducts: RequestHandler = async (req, res) => {
+  const filter = createFilter(req.query);
+
+  try {
+    const result = await getProductsByFilter(filter);
+    return res.status(200).json(result);
+  } catch (error) {
+    return res.status(500).json(error);
+  }
+};
+
+async function getProductsByFilter(filter) {
+  return await Product.findAll({
+    where: filter,
+  });
+}
+
+function createFilter(query: ParsedQs) {
+  const { limited } = query;
+  const { discount } = query;
+  const { rating } = query;
+  const { isNew } = query;
+  const { handpicked } = query;
+  const { minPrice } = query;
+  const { maxPrice } = query;
```

Comment:
Desctructure all of the above using one line:
`const { limited, discount, rating, isNew, handpicked, minPrice, maxPrice } = query;`